packages/api/app/Utils/analytics.js

```diff
@@ -0,0 +1,45 @@
+const { google } = require('googleapis');
+
+const Technology = use('App/Models/Technology');
+
+const Config = use('Adonis/Src/Config');
+
+const { clientEmail, privateKey, viewIds } = Config.get('analytics');
+const scope = ['https://www.googleapis.com/auth/analytics.readonly'];
+
+const authorize = async () => {
+	const jwt = new google.auth.JWT(clientEmail, null, privateKey, scope);
+	await jwt.authorize();
+	return jwt;
+};
+
+const getTechnologyViews = async () => {
+	const jwt = await authorize();
+	const result = await google.analytics('v3').data.ga.get({
+		auth: jwt,
+		ids: `ga:${viewIds.site}`,
+		'start-date': '2021-01-01',
+		'end-date': 'today',
+		metrics: 'ga:pageviews',
+		dimensions: 'ga:pagePath',
+		filters: 'ga:pagePath=~^/t/',
+	});
+	return result.data.rows;
+};
+
+const updateTechnologyTotalViews = async () => {
+	const pageViews = await getTechnologyViews();
+	await Promise.all(
+		pageViews.map(async (pageView) => {
+			const technology = await Technology.findBy('slug', pageView[0].split('/')[2]);
+			technology.total_views = pageView[1];
+			await technology.save();
+		}),
+	);
+};
```

Comment:
Podemos ter um estouro de memória do servidor nessa execução a partir do momento em que tivermos muitas tecnologias a serem atualizadas.

O servidor **sofreria muito** com 10.000 updates no banco rodando paralelamente 😖. E, mesmo que tirarmos o `Promise.all()`, ainda assim o job rodaria por muito tempo.

Podemos pensar em algumas alternativas para isso e conto com sua ajuda pra descobrirmos a melhor implementação pra nosso projeto.

1. Podemos pensar num `update in batch` (ou `bulk update`), fazendo uma única escrita no banco, contendo todo o conjunto de alterações que pretendemos fazer.;
2. Podemos analisar a ideia de escrever num registro do Redis as views para cada tecnologia (ex.: `[{ id: 1, views: 50}, {id: 2, views: 10}]` ) e rodar outro cron job que atualizaria cada registro (podemos ser uma atualização em batch, complementando com o primeiro exemplo);
3. Uma solução menos trabalhosa seria dar um chunk no array `pageViews`, para transformar ele num array de arrays menores, possibilitando a utilização do Promise.all() dentro de algo como `pageViews.forEach(views => Promise.all(views.map(/.../)))`. Dessa maneira simplesmente evitaríamos um estouro de memória 🙂 
3. Desconsiderando os casos acima e fazendo a opção mais simples *(mas não a melhor)*, podemos simplesmente remover o `Promise.all` + `map()` e rodar através de outro laço de repetição qualquer; Aproveitando também pra remover a `select + save` e colocando somente um `update where id...`

O que pensa sobre?

Reply:
Aqui é mostrada uma ideia bem bacana que podemos utilizar: `case` + `when` na query raw. Só não tenho certeza se o MySQL suporta essa sintaxe 👀

Reply:
@mateus4k eu usei a estratégia do CASE WHEN, deu certo!